File: server/src/models/ProductCategory.ts
import {
  Table,
  Model,
  Column,
  IsUUID,
  PrimaryKey,
  Default,
  DataType,
  BelongsTo,
  ForeignKey,
  AllowNull,
} from "sequelize-typescript";
import { Field, ObjectType } from "type-graphql";
import { Product } from "./Product";
import { Category } from "./Category";

@ObjectType()
@Table({
  timestamps: true,
  underscored: true,
})
export class ProductCategory extends Model {
  @Field()
  @IsUUID(4)
  @PrimaryKey
  @Default(DataType.UUIDV4)
  @Column
  id: string;

  @Field()
  @IsUUID(4)
  @ForeignKey(() => Category)
  @AllowNull(false)
  @Column
  categoryId: string;

  @BelongsTo(() => Category)
  category: Category;

  @Field()
  @IsUUID(4)
  @ForeignKey(() => Product)
  @AllowNull(false)
  @Column
  productId: string;

  @BelongsTo(() => Product)
  product: Product;

  @Field()
  createdAt: Date;

  @Field()
  updatedAt: Date;
}

File: go-server/internal/models/product_category.go
package models

func init() {
	registerModel(&ProductCategory{})
}

// ProductCategory is the join table for Product<->Category. Neither
// category/product association nor even categoryId/productId carry @Field()
// in the Node model, so it's never returned directly over GraphQL — it's
// only ever written to via attachCategory/detachCategory.
type ProductCategory struct {
	Base

	CategoryID string    `gorm:"not null;index"`
	Category   *Category `gorm:"foreignKey:CategoryID;constraint:OnDelete:CASCADE"`

	ProductID string   `gorm:"not null;index"`
	Product   *Product `gorm:"foreignKey:ProductID;constraint:OnDelete:CASCADE"`
}
